### PR TITLE
vdf improvements

### PR DESCRIPTION
**Pull request by fchirica**

### CODE REVIEW

File: src/2weso_test.cpp [hunk 1]
```diff
@@ -58,10 +59,11 @@ int main() {
...
     CheckProof(D, proof, iteration);
     // Test 3 - 100 million iters.
-    iteration = 100000000;
+    iteration = 30000000;
     proof = ProveTwoWeso(D, f, iteration, 0, &weso, 0, stopped);
     CheckProof(D, proof, iteration);
     // Test stopping gracefully.
     stopped = true;
     vdf_worker.join();
+    exit(0);
```
Comment (arvidn) [inline src/2weso_test.cpp:68]: the fact that this helps suggests we're corrupting memory somewhere. `-fsanitize=address` might help find it

File: src/threading.h [hunk 2]
```diff
@@ -888,12 +872,9 @@ template<class mpz_type> bool gcd_unsigned(
...
     }
 
-    memory_barrier();
     int error_code=hasAVX2()?
-	    asm_code::asm_avx2_func_gcd_unsigned(&data):
-	    asm_code::asm_cel_func_gcd_unsigned(&data);
-
-    memory_barrier();
+    asm_code::asm_avx2_func_gcd_unsigned(&data):
+    asm_code::asm_cel_func_gcd_unsigned(&data);
```
Comment (arvidn) [inline src/threading.h:877]: I think I messed up indentation here

File: src/vdf_client.cpp [hunk 3]
```diff
@@ -264,18 +264,23 @@ void SessionTwoWeso(tcp::socket& sock) {
 
         // Tell client that I'm ready to get the challenges.
         boost::asio::write(sock, boost::asio::buffer("OK", 2));
-
+        PrintInfo("Written OK.");
```
Comment (arvidn) [inline src/vdf_client.cpp:267]: these `PrintInfo()` should probably be removed before landing

**Comment by arvidn** [comment]
"lazy allocate GMP integers" should be squashed with "save memory by allocating mpz integers lazily" as well as your additional fixes. I think there's an error in your fix though, see comment.

**Comment by wjblanke** [comment]
Hi guys, unfortunately this merge times slower on sumo.chia.net. Let me know if this is not the case on your system.

We should be able to keep the performance the same. Note that the threads move in lockstep so not every variable has to be protected explicitly, in some cases the thread ordering will limit access. Also, on Intel x64 I believe int64s are atomic. Wrapping them with std may be slowing things down? Again just suggestions. While there may be leaks, to my knowledge we haven't had corruption with this code.

(venv) wjblanke@sumo:~$ fc.debug/chiavdf/src/vdf_bench square_asm 4000000
Time: 27760 ms; n_slow: 916; speed: 144.0K ips
a = 0x227d805b2193279654f3eae9fc6653a5a3b5a0307b33d54791f5daf6991eefd63d94fc93d728dcdfc3b9ad8e9d62c0110a2a4a05347f44fee952af6464100d75
b = 0x1a5aa024ffaafca095f851b355bd425de122154feac64aff8da86d902fb01f9dda481055e2722fbb2af55ee46f952c2d405ce90af7a92ce0e03342df225040a1
c = 0x179fde015e2c02d679af11cd68186cbe511377d5d7f2fe86dc3c112c230734341cb8247e0715db552cab8e7a72e17a485fc2a328b8a5f2b3ade19fccaf54dd0de
(venv) wjblanke@sumo:~$ master/chiavdf/src/vdf_bench square_asm 4000000
Time: 27069 ms; n_slow: 916; speed: 147.7K ips
a = 0x227d805b2193279654f3eae9fc6653a5a3b5a0307b33d54791f5daf6991eefd63d94fc93d728dcdfc3b9ad8e9d62c0110a2a4a05347f44fee952af6464100d75
b = 0x1a5aa024ffaafca095f851b355bd425de122154feac64aff8da86d902fb01f9dda481055e2722fbb2af55ee46f952c2d405ce90af7a92ce0e03342df225040a1
c = 0x179fde015e2c02d679af11cd68186cbe511377d5d7f2fe86dc3c112c230734341cb8247e0715db552cab8e7a72e17a485fc2a328b8a5f2b3ade19fccaf54dd0de

**Comment by arvidn** [comment]
it's the x86 `MOV` instruction that's atomic, not naked C++ `int`s. They have to be `std::atomic<int>` in order for the compiler to not re-order them or hoist them in ways where it doesn't matter what the CPU does. Additionally, the compiler may lower certain memory accesses to other instructions than `MOV`, which may not be atomic.

What's a good commit to benchmark against?

**Comment by wjblanke** [comment]
I just compared to master

**Comment by hoffmang9** [comment]
I'm going to close this as I believe we've got it taken care of in the other PRs

**Pull request closed without merging**